libs/hmpb/src/render_ballot.tsx

```diff
@@ -574,25 +605,48 @@ export async function renderAllBallotsAndCreateElectionDefinition<
     JSON.stringify(electionToHash, null, 2)
   ).unsafeUnwrap();
 
-  for (const { document, props } of ballotsWithLayouts) {
-    if (props.ballotMode !== 'sample') {
-      await addQrCodesAndBallotHashes(document, electionDefinition.election, {
-        ballotHash: electionDefinition.ballotHash,
-        ballotStyleId: props.ballotStyleId,
-        precinctId: props.precinctId,
-        ballotType: props.ballotType,
-        isTestMode: props.ballotMode !== 'official',
-        ballotAuditId: props.ballotAuditId,
-      });
-    }
-  }
-
   return {
-    ballotDocuments: ballotsWithLayouts.map((ballot) => ballot.document),
+    ballotContents: ballotLayouts.map(({ ballotContent }) => ballotContent),
     electionDefinition,
   };
 }
 
+export async function renderAllBallotPdfsAndCreateElectionDefinition<
+  P extends BaseBallotProps,
+>(
+  renderer: Renderer,
+  template: BallotPageTemplate<P>,
+  ballotProps: P[],
+  electionSerializationFormat: ElectionSerializationFormat
+): Promise<{
+  ballotPdfs: Uint8Array[];
+  electionDefinition: ElectionDefinition;
+}> {
+  const { ballotContents, electionDefinition } =
+    await layOutBallotsAndCreateElectionDefinition(
+      renderer,
+      template,
+      ballotProps,
+      electionSerializationFormat
+    );
+
+  const ballotPdfs = await iter(ballotProps)
+    .zip(ballotContents)
+    .async()
+    .map(async ([props, ballotContent]) => {
+      const document = await renderer.loadDocumentFromContent(ballotContent);
+      const ballotPdf = await renderBallotPdfWithMetadataQrCode(
+        props,
+        document,
+        electionDefinition
+      );
+      document.cleanup();
```

Comment:
I'm going to ask a silly question to help me grasp what I'm missing..

What would break if we didn't have the `document.cleanup()` and page lock management? It seems like that protects against us rendering two ballots with the same document/page, but I'm missing where there would be any attempt at parallelism. And since there isn't a place we retry after failing to get a lock, I'm not certain the lock capturing is necessary. Sorry if I'm missing obvious but thanks for any clarification

Reply:
This is basically just enforcing the memory optimization changes from this PR for any future usage. Ideally, the API for the renderer would be structured in a way that you could only use it in a memory-efficient way. However, currently, without the lock+cleanup, somebody could come along and write code that tries to use two documents simultaneously (which is how the previous code worked). So you're right that it's not necessary - it's more of a temporary stopgap safeguard until we can redesign the renderer API.

Reply:
Got it, that makes sense to me as the code's written. It's a good safeguard against misuse for now.